File: android/app/src/main/java/app/releaf/mobile/ui/components/AppButton.kt
/*
 * AppButton.kt
 * Primary / secondary (outline) / text button variants.
 */

package app.releaf.mobile.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import app.releaf.mobile.ui.theme.AppColors
import app.releaf.mobile.ui.theme.AppRadius
import app.releaf.mobile.ui.theme.AppSpacing
import app.releaf.mobile.ui.theme.AppTypography

enum class AppButtonVariant { Primary, Secondary, Text }

@Composable
fun AppButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: AppButtonVariant = AppButtonVariant.Primary,
    fillWidth: Boolean = true,
) {
    val background = when (variant) {
        AppButtonVariant.Primary   -> AppColors.ActionPrimary
        AppButtonVariant.Secondary -> AppColors.CardSolid
        AppButtonVariant.Text      -> Color.Transparent
    }
    val foreground = when (variant) {
        AppButtonVariant.Primary   -> AppColors.OnPrimary
        AppButtonVariant.Secondary -> AppColors.TextPrimary
        AppButtonVariant.Text      -> AppColors.Coral
    }
    val borderWidth = if (variant == AppButtonVariant.Secondary) 1.dp else 0.dp
    val borderColor = if (variant == AppButtonVariant.Secondary) AppColors.BorderStrong else Color.Transparent

    val base = modifier
        .clip(RoundedCornerShape(AppRadius.pill))
        .background(background)
        .border(borderWidth, borderColor, RoundedCornerShape(AppRadius.pill))
        .clickable { onClick() }
        .padding(horizontal = AppSpacing.s4, vertical = AppSpacing.s3)

    val layout = if (fillWidth && variant != AppButtonVariant.Text) base.fillMaxWidth() else base

    Row(
        modifier = layout,
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            text = text,
            style = AppTypography.Button,
            color = foreground,
        )
    }
}

File: ios/Releaf/DesignSystem/Components/AppButton.swift
/*
 * AppButton.swift
 * Primary / secondary (outline) / text button variants.
 */

import SwiftUI

public enum AppButtonVariant {
    case primary
    case secondary
    case text
}

public struct AppButton: View {
    public let title: String
    public let variant: AppButtonVariant
    public let icon: Image?
    public let action: () -> Void

    public init(
        _ title: String,
        variant: AppButtonVariant = .primary,
        icon: Image? = nil,
        action: @escaping () -> Void
    ) {
        self.title = title
        self.variant = variant
        self.icon = icon
        self.action = action
    }

    public var body: some View {
        Button(action: action) {
            HStack(spacing: AppSpacing.s2) {
                if let icon = icon {
                    icon
                        .renderingMode(.template)
                        .resizable()
                        .frame(width: 16, height: 16)
                }
                Text(title)
                    .font(AppText.button)
            }
            .padding(.horizontal, AppSpacing.s6)
            .padding(.vertical, AppSpacing.s3)
            .frame(maxWidth: variant == .text ? nil : .infinity)
            .foregroundStyle(foreground)
            .background(background)
            .overlay(
                RoundedRectangle(cornerRadius: AppRadius.lg)
                    .stroke(borderColor, lineWidth: borderWidth)
            )
            .clipShape(RoundedRectangle(cornerRadius: AppRadius.lg))
        }
        .buttonStyle(.plain)
    }

    private var foreground: Color {
        switch variant {
        case .primary:   return AppColors.onPrimary
        case .secondary: return AppColors.textPrimary
        case .text:      return AppColors.coral
        }
    }

    private var background: Color {
        switch variant {
        case .primary:   return AppColors.actionPrimary
        case .secondary: return AppColors.cardSolid
        case .text:      return .clear
        }
    }

    private var borderColor: Color {
        switch variant {
        case .primary:   return .clear
        case .secondary: return AppColors.borderStrong
        case .text:      return .clear
        }
    }

    private var borderWidth: CGFloat {
        switch variant {
        case .secondary: return 1
        default:         return 0
        }
    }
}

#Preview {
    ZStack {
        AppColors.canvas.ignoresSafeArea()
        VStack(spacing: AppSpacing.s3) {
            AppButton("Continue") {}
            AppButton("Cancel", variant: .secondary) {}
            AppButton("Learn more", variant: .text) {}
        }
        .padding()
    }
}
